### DISCUSSION TITLE
Older PC4020 panel

### DISCUSSION THREAD

**Discussion by PapaPhred** [taligentx/dscKeybusInterface#172] [dscKeybusInterface Discussion]
I have been watching this project with great interest and I’m getting close to jumping into the pool ;-). I have just obtained a second newer firmware 4020 panel to allow testing, but I know there are some differences between my old MaxSys PC4020 and the “Power Series” boards. Ie. I have to use specific keyboards with this panel etc.  
can anyone tell me if I might expect any success with this code and this older model and are the differences physical (as in voltage levels etc.) or mainly protocol differences. 
I do have a couple of logic analyzers that I can use to map out the command/response values, but is it possible that this setup will be able to interface with the com buss on this model at all???
Thanks!!

**Comment by taligentx** [comment]
Hi @PapaPhred - I wasn't aware of the MaxSys series until you mentioned it, but it does seem to share some accessories with the PowerSeries:

https://www.dsc.com/alarm-security-products/TL250%20-%20Internet%20Alarm%20Communicator%20-%20PowerSeries/Maxsys/140

So the protocol and voltages may be similar, it'd be interesting to see the output from the `KeybusReader` sketch for this panel if you're willing to experiment!

**Comment by kricon** [comment]
@taligentx tlink TL250 communicator module connects with PC-Link to the panel, not thru keybus.
There are no modules which connects thru keybus with MaxSys and PowerSeries. Even wireless expanders are different (pc5132 for power, pc4164 for maxsys, also the pc5108 for power and pc4108/pc4116 for maxsys).
Keybus on TL250 communicator is for local keypad programming thru LCD5500 and for PC5108 input expander when more inputs than onboard 4 are needed. I think that TL250 IP Communicator can even work standalone, without need for a panel.
Tlink TL300 connects to ring/tip telephone dialer on the panel and simulates a telephone line, TL250 connects thru pc-link.

I suppose that keybus on maxsys works on same voltages and duty cycle, but that commands are completely different. That is just what I think - I cant prove it as I newer had hands on any Maxsys panel. But recently I obtain few modules like Tlink TL250, Maxsys PC4116 16-zone expansion module and LCD4501 maxsys keyboards which I tested on Powerseries and it gives "System/keybus trouble, # to silence". I can run `KeybusReader` on powerseries with any of the modules I have if needed.

**Comment by PapaPhred** [comment]
I’m hopeful that at least the primary functions will work or be very close ;-). If nothing else that it will be able to read and write the Com bus. I still have a few more things to get off my plate before I can dedicate the required cycles to this, but I have a spare panel and a few keypads and a couple of clone logic analyzers (USBEE-AX and Salea) which should do the job.  Once I have things setup and start capturing some command/response values I will start posting these.  It’s a quite capable panel (16 zones on panel expandable to 128, 16 keypads, 8 partitions and up to 64 wireless zones etc. ). It was pretty bad a$$ when it first came out ;-). I will do my best to map as much as I can (don’t have most of the expansion modules at this time) and hopefully it will be useful to another poor soul like me hanging on to a relic ;-)
I want to say thanks in advance for all the effort that has been made to make this possible!!!!

Phred Broughton
        Ô¿Ô¬

**Comment by PapaPhred** [comment]
Meant to include this ;-)

https://www.dsc.com/alarm-security-products/PC4020%20-%20MAXSYS%20Control%20Panel/4


Phred Broughton
        Ô¿Ô¬

**Comment by ekrotki** [comment]
Hi, I do have too MaxSys PC4020A located in remote location, installed in 2000 (yes, over 20 years ago). 

Installation has two LCD panels (donwstairs and upstairs) connected via "combus" communication lines with the central unit.
I don't have any hardware-level documentation nor logic analyzers handy so I can only rely on results of your investigations.

My experience is, so far, negative, but I still have hope. 

Just installed ESP8266 with telegram integration and found it not working with "combus" interface, connected directly to central unit panel red, black, yellow, green lines. Yes, 33k/10k interface is used, with BC547 transistor for "writing". 

It seems to me that PC4020 unit's "combus" is different than the "keybus" of the newer units you interface with. 
I am living over 700 km far from this place, so any direct interaction with hardware is rarely possible, this is why I implemented OTA with telegram so I can deploy software fixes assuming it will not break OTA or telegram integration :).

I would kindly appreciate any guidance for me to get this working with DSC.

Thanks in advance
Emil

**Comment by Dilbert66** [comment]
It would take some logic analyzer traces to analyze this board. There is no way to get it working without understanding how this  bus differs from the others and since it seems no one with the capability to analyze it has a sample board to test with, I can't see this being happening any time soon.

**Comment by PapaPhred** [comment]
I have the PC4020 installed at my home with 3 keypads connected. At the time I installed it I had to find a couple of keypads and do remember having to find the correct keypads because the newer  units wouldn’t work with the older “MaxSys” combos protocol. It seems (and I am hoping) that the changes were more additions than revisions which may mean that we could re-use some of what has already been done.  I suspect it will be affected by firmware version as several modules are only compatible with later versions. Guess that remains to be seen 😊

cannot make a promise yet, but I have picked up a second PC4020 and I have been hoping to spend some time with my logic analyzer to get this working. I will check back and let you know as I get closer to this. Hopefully it won’t be much longer before I can get started! 

BTW, my greatest thanks to those who have gone before and gotten this working so well

Hoping to get some time very soon!!

 

 

 

Phred Broughton

        Ô¿Ô¬

**Comment by Dilbert66** [comment]
A good start is to just post saved traces of the yellow and green lines using Pulseview or Saelae logic software and we can then at least get an idea what is going on. We can just use your traces in our our own software copies to try and determine the protocol that way.  My personal favorite is Pulseview as a logic analyzer.

**Comment by PapaPhred** [comment]
I have two boards. My original board is a very early firmware 😊 like version 1.0, but the board I just bought is much later version 3.x (not home now). I noticed that the IP add-in board (think it’s an escort something) which lists as incompatible with the version 1 is supported in 3.x so seems there were some useful updates. I don’t think it will grossly change the “protocol” other than to support new items 😉

 

I’ll let you know when I get things moving.

Thanks for keeping this going!!!

 

 

 

 

Phred Broughton

        Ô¿Ô¬

**Comment by Dilbert66** [comment]
Have you tried the keybusreader script that was suggested earlier? It would also provide insight into what the protocol is assuming that the basic hardware communications is similar.  It might just be the commands that differ.

**Comment by PapaPhred** [comment]
That’s what I am hoping to find.  Been trying to get some time to dig in to this for a while! Haven’t actually even connected the analyzer to the panel yet, but I will give the script a try first for sure!!

Phred Broughton
        Ô¿Ô¬

**Comment by ekrotki** [comment]
I am really sad I can't participate in this investigation for two reasons - I don't even have an oscilloscope and second - I am back at work, 700 km far from the installation....

**Comment by Dilbert66** [comment]
Here's a little help on setting up Pulseview to analyze a DSC panel. 

-  D0 on the yellow line (connect it after the 33k/10k resistor divider to keep the pulse voltate below 5volt so as not to damage the analyzer inputs) and D1 to the green line
- connect your saelae (or clone) to the usb port of your computer and startup Pulseview. It should find your logic analyzer.
- click run to collect some samples.  If you see data, you can now setup the decoder.  It should look similar to this for d0/d1 (ignore the panel/keypad decoder lines for now). You need to use a decoder to analyze it properly due to the signal multiplexing.

![dsc_capture](https://user-images.githubusercontent.com/7193213/136216017-8a7e551b-985b-4992-9d56-8316225e908b.png)

The DSC protocol uses a clock (with syncronizing pulses) on the yellow line and the green line is used for bidirectional data where the panel sends on the rising clock edge and peripherals send on the falling clock edge.  The sync pulse is a high signal of about 5.5 ms of duration. This signals a new cmd. On the next rising edge of the clock after the sync, the panel starts sending it's cmd data.  Any device needing to respond will send it's data after the 9th panel bit  on the falling edges of the clock.

To decode this we need a synchronous decoder that allows us to set  a clock and data channel and set the trigger  on the clock polarity or edge.  I used the pulseview SPI decoder as it had the functions needed and set it up like shown below:

![spi_decoder_setup](https://user-images.githubusercontent.com/7193213/136216061-f028787b-4fa6-40a4-96ed-c4ff9bd78fc7.png)

This is for the Panel setup.  clk is d0, miso is d1 cs# polarity is active-high. Clock polarity is 0.

For the Keypad decoder setup, the only change is to set the clock polarity to 1.  This means that it will use the inverted clock pulse (ie similar to a falling edge to trigger on).

As you can see in the dsc capture above you can view the 00000101 0 10011000 which is a 05 cmd which are sent constantly.  

Note:  We are not actually decoding the data as this would require a custom decoder written, I am simply using the SPI decoder as a hack to split the keypad/panel data streams into their own channels for analysis.

For the pc4020, it might all be completely different and this is what we need to determine.   

Hope this helps in your research.

Alain

**Reply by taligentx** [reply]
Alain, thanks for this excellent writeup!

**Comment by PapaPhred** [comment]
Many thanks!!!
Hopefully it wont be too long before I get to this!!!

Phred Broughton
        Ô¿Ô¬
